crates/proto/src/serialize/binary/encoder.rs

```diff
@@ -472,6 +490,34 @@ pub enum NameEncoding {
     UncompressedLowercase,
 }
 
+/// This wraps a [BinEncoder] and applies different name encoding options.
+///
+/// Original name encoding options will be restored when this is dropped.
+pub struct ModalEncoder<'a, 'e> {
```

Comment:
Nit: can we fold this to only use a single lifetime?

Reply:
No, if there's only one lifetime parameter, then I get "lifetime may not live long enough" errors at call sites. We need to be able to express that the encoder's byte slice outlives the method call, while the reborrowed mutable reference to the encoder may not.